services/hanke-service/src/integrationTest/kotlin/fi/hel/haitaton/hanke/permissions/HankeKayttajaControllerITest.kt

```diff
@@ -0,0 +1,124 @@
+@Import(IntegrationTestConfiguration::class)
+@ActiveProfiles("itest")
+@WithMockUser(USERNAME)
+class HankeKayttajaControllerITest(@Autowired override val mockMvc: MockMvc) : ControllerTest {
+
+    @Autowired private lateinit var hankeKayttajaService: HankeKayttajaService
+    @Autowired private lateinit var hankeService: HankeService
+    @Autowired private lateinit var permissionService: PermissionService
+    @Autowired private lateinit var disclosureLogService: DisclosureLogService
+
+    @BeforeEach
+    fun clearMocks() {
+        clearAllMocks()
+    }
+
+    @AfterEach
+    fun checkMocks() {
+        checkUnnecessaryStub()
+        confirmVerified(hankeKayttajaService, hankeService, permissionService)
+    }
+
+    @Test
+    fun `getHankeKayttajat when valid request returns users of given hanke and logs audit`() {
+        val hanke = HankeFactory.create()
+        val testData = generateContacts()
+        every { hankeService.findHankeOrThrow(HANKE_TUNNUS) } returns hanke
+        justRun { permissionService.verifyHankeUserAuthorization(USERNAME, hanke, VIEW) }
+        every { hankeKayttajaService.getKayttajatByHankeId(hanke.id!!) } returns testData
+
+        val response: HankeKayttajaResponse =
+            getHankeKayttajat().andExpect(status().isOk).andReturnBody()
+
+        assertThat(response.kayttajat).hasSize(3)
+        with(response.kayttajat.first()) {
+            assertThat(id).isNotNull()
+            assertThat(nimi).isEqualTo("test name1")
+            assertThat(sahkoposti).isEqualTo("email.1.address.com")
+            assertThat(tunnistautunut).isEqualTo(false)
+        }
+        assertThat(response.kayttajat).hasSameElementsAs(testData)
+        verifyOrder {
+            hankeService.findHankeOrThrow(HANKE_TUNNUS)
+            permissionService.verifyHankeUserAuthorization(USERNAME, hanke, VIEW)
+            hankeKayttajaService.getKayttajatByHankeId(hanke.id!!)
+            disclosureLogService.saveDisclosureLogsForHankeKayttajat(response.kayttajat, USERNAME)
+        }
+    }
+
+    @Test
+    fun `getHankeKayttajat when no permission for hanke returns not found`() {
+        val hanke = HankeFactory.create()
+        every { hankeService.findHankeOrThrow(HANKE_TUNNUS) } returns hanke
+        every { permissionService.verifyHankeUserAuthorization(USERNAME, hanke, VIEW) } throws
+            HankeNotFoundException(HANKE_TUNNUS)
+
+        getHankeKayttajat().andExpect(status().isNotFound)
+
+        verifyOrder {
+            hankeService.findHankeOrThrow(HANKE_TUNNUS)
+            permissionService.verifyHankeUserAuthorization(USERNAME, hanke, VIEW)
+        }
+        verify { hankeKayttajaService wasNot Called }
+    }
+
+    @Test
+    @WithAnonymousUser
+    fun `getHankeKayttajat when unauthorized token returns 401 `() {
+        getHankeKayttajat().andExpect(status().isUnauthorized)
+    }
+
+    private fun generateContacts(amount: Int = 3): List<HankeKayttajaDto> =
```

Comment:
Isn't this basically the same method as `createMockHankeKayttajat` in `DisclosureLogServiceTest`? Make this a factory method?

Reply:
Thanks, they are duplicate. Moved to a factory object.